packages/trackerless-network/src/logic/StreamrNode.ts

```diff
@@ -0,0 +1,404 @@
+/* eslint-disable @typescript-eslint/parameter-properties */
+
+import { RandomGraphNode } from './RandomGraphNode'
+import {
+    PeerDescriptor,
+    ConnectionLocker,
+    DhtNode,
+    ITransport,
+    keyFromPeerDescriptor
+} from '@streamr/dht'
+import { StreamMessage } from '../proto/packages/trackerless-network/protos/NetworkRpc'
+import { EventEmitter } from 'eventemitter3'
+import {
+    Logger,
+    MetricsContext,
+    RateMetric,
+    Metric,
+    MetricsDefinition,
+    waitForEvent3
+} from '@streamr/utils'
+import { uniq } from 'lodash'
+import { StreamPartID, StreamPartIDUtils } from '@streamr/protocol'
+import { sampleSize } from 'lodash'
+import { StreamEntryPointDiscovery } from './StreamEntryPointDiscovery'
+import { ILayer0 } from './ILayer0'
+import { createRandomGraphNode } from './createRandomGraphNode'
+import { ProxyDirection } from '../proto/packages/trackerless-network/protos/NetworkRpc'
+import { IStreamNode } from './IStreamNode'
+import { ProxyStreamConnectionClient } from './proxy/ProxyStreamConnectionClient'
+import { PeerIDKey } from '@streamr/dht/src/exports'
+
+export enum StreamNodeType {
+    RANDOM_GRAPH = 'random-graph',
+    PROXY = 'proxy'
+}
+
+export interface NeighborCounterEvents {
+    targetReached: () => void
+}
+
+class NeighborCounter {
+    private counter = 0
+    private emitter = new EventEmitter<NeighborCounterEvents>()
+
+    constructor(private randomGraphNode: RandomGraphNode, private targetNumberOfNeighbors: number) {
+        this.counter = randomGraphNode.getTargetNeighborStringIds().length
+        randomGraphNode.on('targetNeighborConnected', this.onTargetNeighborConnected)
+    }
+
+    private onTargetNeighborConnected = (_id: string) => {
+        this.counter++
+        if (this.counter == this.targetNumberOfNeighbors) {
+            this.randomGraphNode.off('targetNeighborConnected', this.onTargetNeighborConnected)
+            this.emitter.emit('targetReached')
+        }
+    }
+
+    public async waitForTargetReached(timeout: number): Promise<void> {
+        if (this.counter >= this.targetNumberOfNeighbors) {
+            return
+        } else {
+            await waitForEvent3<NeighborCounterEvents>(this.emitter, 'targetReached', timeout)
+        }
+    }
+}
+
+export interface StreamObject {
+    layer1?: DhtNode
+    layer2: IStreamNode
+    type: StreamNodeType
+}
+
+export interface Events {
+    newMessage: (msg: StreamMessage) => void
+}
+
+const logger = new Logger(module)
+
+let cleanUp: () => Promise<void> = async () => { }
+
+interface Metrics extends MetricsDefinition {
+    publishMessagesPerSecond: Metric
+    publishBytesPerSecond: Metric
+}
+
+export interface StreamrNodeConfig {
+    metricsContext?: MetricsContext
+    id?: string
+    streamPartitionNumOfNeighbors?: number
+    streamPartitionMinPropagationTargets?: number
+    nodeName?: string
```

Comment:
Do we need both name and id? What is the difference of those fields? If we need both, we could rename `nodeName` to name` `as all of these are related to a node.

Reply:
Name is a human-readable name given for the nodes to make debugging easier. The ids are not passed in a human-readable format in the network

Reply:
But would it be possible to pass human-readable ids as node ids? What requirements are there for the id: it maybe needs to be globally unique, but are the other limitations?